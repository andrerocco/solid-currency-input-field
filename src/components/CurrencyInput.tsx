import type { JSX } from 'solid-js'
import { createSignal, createEffect, createMemo, mergeProps, splitProps } from 'solid-js'
import { CurrencyInputProps, CurrencyInputOnChangeValues } from './CurrencyInputProps'
import {
  isNumber,
  cleanValue,
  fixedDecimalValue,
  formatValue,
  getLocaleConfig,
  padTrimValue,
  CleanValueOptions,
  getSuffix,
  FormatValueOptions,
  repositionCursor,
} from './utils'

export const CurrencyInput = (
  /* {
   *   allowDecimals = true,
   *   allowNegativeValue = true,
   *   id,
   *   name,
   *   className,
   *   customInput,
   *   decimalsLimit,
   *   defaultValue,
   *   disabled = false,
   *   maxLength: userMaxLength,
   *   value: userValue,
   *   onValueChange,
   *   fixedDecimalLength,
   *   placeholder,
   *   decimalScale,
   *   prefix,
   *   suffix,
   *   intlConfig,
   *   step,
   *   min,
   *   max,
   *   disableGroupSeparators = false,
   *   disableAbbreviations = false,
   *   decimalSeparator: _decimalSeparator,
   *   groupSeparator: _groupSeparator,
   *   onChange,
   *   onFocus,
   *   onBlur,
   *   onKeyDown,
   *   onKeyUp,
   *   transformRawValue,
   *   ...props
   * } */
  baseProps: CurrencyInputProps,
) => {
  const mergedProps = mergeProps(
    {
      allowDecimals: true,
      allowNegativeValue: true,
      disabled: false,
      disableGroupSeparators: false,
      disableAbbreviations: false,
    },
    baseProps,
  )
  const [props, HTMLInputProps] = splitProps(mergedProps, [
    'transformRawValue',
    'onKeyUp',
    'onKeyDown',
    'onBlur',
    'onFocus',
    'onChange',
    'groupSeparator',
    'decimalSeparator',
    'disableAbbreviations',
    'disableGroupSeparators',
    'max',
    'min',
    'step',
    'intlConfig',
    'suffix',
    'prefix',
    'decimalScale',
    'placeholder',
    'fixedDecimalLength',
    'onValueChange',
    'value',
    'maxLength',
    'disabled',
    'defaultValue',
    'decimalsLimit',
    'customInput',
    'class',
    'name',
    'id',
    'allowNegativeValue',
    'allowDecimals',
  ])
  createEffect(() => {
    if (props.decimalSeparator && isNumber(props.decimalSeparator)) {
      throw new Error('decimalSeparator cannot be a number')
    }

    if (props.groupSeparator && isNumber(props.groupSeparator)) {
      throw new Error('groupSeparator cannot be a number')
    }
  })

  const localeConfig = createMemo(() => getLocaleConfig(props.intlConfig))
  const decimalSeparator = createMemo(
    () => props.decimalSeparator || localeConfig().decimalSeparator || '',
  )
  const groupSeparator = createMemo(
    () => props.groupSeparator || localeConfig().groupSeparator || '',
  )

  createEffect(() => {
    if (
      decimalSeparator() &&
      groupSeparator() &&
      decimalSeparator() === groupSeparator() &&
      !props.disableGroupSeparators
    ) {
      throw new Error('decimalSeparator cannot be the same as groupSeparator')
    }
  })

  const formatValueOptions = createMemo(
    (): Partial<FormatValueOptions> => ({
      decimalSeparator: decimalSeparator(),
      groupSeparator: groupSeparator(),
      disableGroupSeparators: props.disableGroupSeparators,
      intlConfig: props.intlConfig,
      prefix: props.prefix || localeConfig().prefix,
      suffix: props.suffix,
    }),
  )

  const cleanValueOptions = (): Partial<CleanValueOptions> => ({
    decimalSeparator: decimalSeparator(),
    groupSeparator: groupSeparator(),
    allowDecimals: props.allowDecimals,
    decimalsLimit: props.decimalsLimit || props.fixedDecimalLength || 2,
    allowNegativeValue: props.allowNegativeValue,
    disableAbbreviations: props.disableAbbreviations,
    prefix: props.prefix || localeConfig().prefix,
    transformRawValue: props.transformRawValue,
  })

  const formattedStateValue = () =>
    props.defaultValue !== undefined
      ? formatValue({
          ...formatValueOptions(),
          decimalScale: props.decimalScale,
          value: String(props.defaultValue),
        })
      : props.value !== undefined
      ? formatValue({
          ...formatValueOptions(),
          decimalScale: props.decimalScale,
          value: String(props.value),
        })
      : ''

  // eslint-disable-next-line solid/reactivity
  const [stateValue, setStateValue] = createSignal(formattedStateValue())
  const [dirty, setDirty] = createSignal(false)
  const [cursor, setCursor] = createSignal(0)
  const [changeCount, setChangeCount] = createSignal(0)
  const [lastKeyStroke, setLastKeyStroke] = createSignal<string | null>(null)
  let inputRef: HTMLInputElement | undefined
  /* useImperativeHandle(props.ref, () => inputRef); */

  /**
   * Process change in value
   */
  const processChange = (value: string, selectionStart?: number | null): void => {
    setDirty(true)

    const { modifiedValue, cursorPosition } = repositionCursor({
      selectionStart,
      value,
      lastKeyStroke: lastKeyStroke(),
      stateValue: stateValue(),
      groupSeparator: groupSeparator(),
    })

    const stringValue = cleanValue({ value: modifiedValue, ...cleanValueOptions() })

    if (props.maxLength && stringValue.replace(/-/g, '').length > props.maxLength) {
      return
    }

    if (stringValue === '' || stringValue === '-' || stringValue === decimalSeparator()) {
      props.onValueChange &&
        props.onValueChange(undefined, props.name, { float: undefined, formatted: '', value: '' })
      setStateValue(stringValue)
      // Always sets cursor after '-' or decimalSeparator input
      setCursor(1)
      return
    }

    const stringValueWithoutSeparator = decimalSeparator()
      ? stringValue.replace(decimalSeparator(), '.')
      : stringValue

    const numberValue = parseFloat(stringValueWithoutSeparator)

    const formattedValue = formatValue({
      value: stringValue,
      ...formatValueOptions(),
    })

    if (cursorPosition !== undefined && cursorPosition !== null) {
      // Prevent cursor jumping
      let newCursor = cursorPosition + (formattedValue.length - value.length)
      newCursor = newCursor <= 0 ? (props.prefix ? props.prefix.length : 0) : newCursor

      setCursor(newCursor)
      setChangeCount(count => count + 1)
    }

    setStateValue(formattedValue)

    if (props.onValueChange) {
      const values: CurrencyInputOnChangeValues = {
        float: numberValue,
        formatted: formattedValue,
        value: stringValue,
      }
      props.onValueChange(stringValue, props.name, values)
    }
  }

  /**
   * Handle change event
   */
  const handleOnChange: JSX.ChangeEventHandler<HTMLInputElement, Event> = event => {
    const {
      target: { value, selectionStart },
    } = event

    processChange(value, selectionStart)
    event.target.value = getRenderValue()
    event.target.setSelectionRange(cursor(), cursor())
    props.onChange && props.onChange(event)
  }

  /**
   * Handle focus event
   */
  const handleOnFocus: JSX.FocusEventHandler<HTMLInputElement, FocusEvent> = event => {
    props.onFocus && props.onFocus(event)
    return stateValue() ? stateValue().length : 0
  }

  /**
   * Handle blur event
   *
   * Format value by padding/trimming decimals if required by
   */
  const handleOnBlur: JSX.FocusEventHandler<HTMLInputElement, FocusEvent> = event => {
    const {
      target: { value },
    } = event

    const valueOnly = cleanValue({ value, ...cleanValueOptions() })

    if (valueOnly === '-' || valueOnly === decimalSeparator() || !valueOnly) {
      setStateValue('')
      props.onBlur && props.onBlur(event)
      return
    }

    const fixedDecimals = fixedDecimalValue(valueOnly, decimalSeparator(), props.fixedDecimalLength)

    const newValue = padTrimValue(
      fixedDecimals,
      decimalSeparator(),
      props.decimalScale !== undefined ? props.decimalScale : props.fixedDecimalLength,
    )

    const numberValue = parseFloat(newValue.replace(decimalSeparator(), '.'))

    const formattedValue = formatValue({
      ...formatValueOptions(),
      value: newValue,
    })

    if (props.onValueChange) {
      props.onValueChange(newValue, props.name, {
        float: numberValue,
        formatted: formattedValue,
        value: newValue,
      })
    }

    setStateValue(formattedValue)

    props.onBlur && props.onBlur(event)
  }

  /**
   * Handle key down event
   *
   * Increase or decrease value by step
   */
  const handleOnKeyDown: JSX.EventHandler<HTMLInputElement, KeyboardEvent> = event => {
    const { key } = event

    setLastKeyStroke(key)

    if (props.step && (key === 'ArrowUp' || key === 'ArrowDown')) {
      event.preventDefault()
      setCursor(stateValue().length)

      const currentValue =
        parseFloat(
          props.value !== undefined
            ? String(props.value).replace(decimalSeparator(), '.')
            : cleanValue({ value: stateValue(), ...cleanValueOptions() }),
        ) || 0
      const newValue = key === 'ArrowUp' ? currentValue + props.step : currentValue - props.step

      if (props.min !== undefined && newValue < props.min) {
        return
      }

      if (props.max !== undefined && newValue > props.max) {
        return
      }

      const fixedLength = String(props.step).includes('.')
        ? Number(String(props.step).split('.')[1]?.length)
        : undefined

      processChange(
        String(fixedLength ? newValue.toFixed(fixedLength) : newValue).replace(
          '.',
          decimalSeparator(),
        ),
      )
    }

    props.onKeyDown && props.onKeyDown(event)
  }

  /**
   * Handle key up event
   *
   * Move cursor if there is a suffix to prevent user typing past suffix
   */
  const handleOnKeyUp: JSX.EventHandler<HTMLInputElement, KeyboardEvent> = event => {
    const {
      key,
      currentTarget: { selectionStart },
    } = event
    if (key !== 'ArrowUp' && key !== 'ArrowDown' && stateValue() !== '-') {
      const suffix = getSuffix(stateValue(), {
        groupSeparator: groupSeparator(),
        decimalSeparator: decimalSeparator(),
      })

      if (suffix && selectionStart && selectionStart > stateValue().length - suffix.length) {
        /* istanbul ignore else */
        if (inputRef) {
          const newCursor = stateValue().length - suffix.length
          inputRef.setSelectionRange(newCursor, newCursor)
        }
      }
    }

    props.onKeyUp && props.onKeyUp(event)
  }

  createEffect(() => {
    // prevent cursor jumping if editing value
    changeCount()
    if (dirty() && stateValue() !== '-' && inputRef && document.activeElement === inputRef) {
      inputRef.setSelectionRange(cursor(), cursor())
    }
  })

  /**
   * If user has only entered "-" or decimal separator,
   * keep the char to allow them to enter next value
   */
  const getRenderValue = createMemo(() => {
    if (
      props.value !== undefined &&
      stateValue() !== '-' &&
      (!decimalSeparator() || stateValue() !== decimalSeparator())
    ) {
      return formatValue({
        ...formatValueOptions(),
        decimalScale: dirty() ? undefined : props.decimalScale,
        value: String(props.value),
      })
    }
    return stateValue()
  })

  const inputProps = (): JSX.InputHTMLAttributes<HTMLInputElement> => ({
    ...HTMLInputProps,
    type: 'text',
    id: props.id,
    name: props.name,
    class: props.class,
    inputMode: 'decimal' as const,
    onInput: handleOnChange,
    onBlur: handleOnBlur,
    onFocus: handleOnFocus,
    onKeyDown: handleOnKeyDown,
    onKeyUp: handleOnKeyUp,
    placeholder: props.placeholder,
    disabled: props.disabled,
    value: getRenderValue(),
  })

  /* if (props.customInput) {
   *   const CustomInput = props.customInput;
   *   return <CustomInput {...inputProps} />;
   * } */

  return <input {...inputProps()} ref={inputRef} />
}

export default CurrencyInput
