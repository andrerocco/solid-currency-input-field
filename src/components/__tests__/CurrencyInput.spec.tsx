import '@testing-library/jest-dom'
import { render, fireEvent, screen } from '@solidjs/testing-library'
import userEvent from '@testing-library/user-event'
import CurrencyInput from '../CurrencyInput'
import { vi } from 'vitest'

describe('<CurrencyInput/>', () => {
  const onValueChangeSpy = vi.fn()

  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('Renders without error', () => {
    render(() => (
      <CurrencyInput
        id="validationCustom01"
        name="inputName"
        placeholder="£1,000"
        className="form-control"
      />
    ))
    const input = screen.getByRole('textbox')

    expect(input).toMatchSnapshot()

    expect(input).toHaveValue('')
  })

  it('Renders with default value', () => {
    render(() => <CurrencyInput defaultValue={1234.56} prefix="£" />)
    const input = screen.getByRole('textbox')

    expect(input).toMatchSnapshot()

    expect(input).toHaveValue('£1,234.56')
  })

  it('Renders with default value 0', () => {
    render(() => <CurrencyInput defaultValue={0} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('£0')
  })

  it('Renders with default value undefined', () => {
    render(() => <CurrencyInput defaultValue={undefined} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('Renders with default value null', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    render(() => <CurrencyInput defaultValue={null as any} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('Renders with default value 0 with decimalScale 2', () => {
    render(() => <CurrencyInput defaultValue={0} decimalScale={2} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('£0.00')
  })

  it('Renders with value prop', () => {
    render(() => <CurrencyInput value={49.99} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('£49.99')
  })

  it('Renders with value 0', () => {
    render(() => <CurrencyInput value={0} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('£0')
  })

  it('Renders with value undefined', () => {
    render(() => <CurrencyInput value={undefined} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('Renders with value null', () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    render(() => <CurrencyInput value={null as any} prefix="£" />)

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('Renders with value 0 with decimalScale 2', () => {
    render(() => <CurrencyInput value={0} prefix="£" decimalScale={2} />)

    expect(screen.getByRole('textbox')).toHaveValue('£0.00')
  })

  it('Renders with value 0.1 with decimalScale 2', async () => {
    render(() => <CurrencyInput value={0.1} prefix="£" decimalScale={2} />)

    expect(screen.getByRole('textbox')).toHaveValue('£0.10')

    await userEvent.type(screen.getByRole('textbox'), '{backspace}')

    expect(screen.getByRole('textbox')).toHaveValue('£0.1')
  })

  it('should go to end of string on focus', async () => {
    render(() => <CurrencyInput defaultValue={123} />)
    await userEvent.type(screen.getByRole('textbox'), '{arrowleft}4{arrowright}6')

    expect(screen.getByRole('textbox')).toHaveValue('12,436')
  })

  it('should allow value change with number', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '100')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('100', undefined, {
      float: 100,
      formatted: '£100',
      value: '100',
    })
  })

  it('should prefix 0 value', () => {
    render(() => <CurrencyInput prefix="£" value={0} onValueChange={onValueChangeSpy} />)

    expect(screen.getByRole('textbox')).toHaveValue('£0')
  })

  it('should allow 0 value on change', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '0')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('0', undefined, {
      float: 0,
      formatted: '£0',
      value: '0',
    })

    expect(screen.getByRole('textbox')).toHaveValue('£0')
  })

  it('should allow empty value', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} defaultValue={1} />)
    await userEvent.clear(screen.getByRole('textbox'))

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    })

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('should callback name as second parameter if name prop provided', async () => {
    const name = 'inputName'
    render(() => <CurrencyInput name={name} prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '123')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('123', name, {
      float: 123,
      formatted: '£123',
      value: '123',
    })
  })

  it('should not allow invalid characters', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), 'hello')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    })

    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('should ignore invalid characters', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '£123hello')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('123', undefined, {
      float: 123,
      formatted: '£123',
      value: '123',
    })

    expect(screen.getByRole('textbox')).toHaveValue('£123')
  })

  it('should clear decimal point only input', async () => {
    render(() => (
      <>
        {' '}
        <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />
        <button data-testid="button" />
      </>
    ))
    await userEvent.type(screen.getByRole('textbox'), '.')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith(undefined, undefined, {
      float: null,
      formatted: '',
      value: '',
    })

    await userEvent.click(screen.getByTestId('button'))
    expect(screen.getByRole('textbox')).toHaveValue('')
  })

  it('should allow .3 decimal inputs', async () => {
    render(() => <CurrencyInput prefix="£" onValueChange={onValueChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '.3')

    expect(onValueChangeSpy).toHaveBeenLastCalledWith('.3', undefined, {
      float: 0.3,
      formatted: '£0.3',
      value: '.3',
    })

    fireEvent.focusOut(screen.getByRole('textbox'))
    expect(screen.getByRole('textbox')).toHaveValue('£0.3')
  })

  it('should call onChange', async () => {
    const onChangeSpy = vi.fn()
    render(() => <CurrencyInput prefix="£" onChange={onChangeSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '123')

    expect(onChangeSpy).toBeCalledTimes(3)

    expect(screen.getByRole('textbox')).toHaveValue('£123')
  })

  it('should call onBlur', async () => {
    const onBlurSpy = vi.fn()
    render(() => (
      <>
        <CurrencyInput prefix="£" onBlur={onBlurSpy} />
        <button data-testid="button" />
      </>
    ))
    await userEvent.click(screen.getByRole('textbox'))
    await userEvent.click(screen.getByTestId('button'))

    expect(onBlurSpy).toBeCalledTimes(1)
  })

  it('should call onFocus', async () => {
    const onFocusSpy = vi.fn()
    render(() => <CurrencyInput onFocus={onFocusSpy} />)
    await userEvent.click(screen.getByRole('textbox'))

    expect(onFocusSpy).toBeCalledTimes(1)
  })

  it('should call onKeyDown', async () => {
    const onKeyDownSpy = vi.fn()
    render(() => <CurrencyInput onKeyDown={onKeyDownSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '1')

    expect(onKeyDownSpy).toBeCalledTimes(1)
  })

  it('should call onKeyUp', async () => {
    const onKeyUpSpy = vi.fn()
    render(() => <CurrencyInput onKeyUp={onKeyUpSpy} />)
    await userEvent.type(screen.getByRole('textbox'), '1')

    expect(onKeyUpSpy).toBeCalledTimes(1)
  })
})
