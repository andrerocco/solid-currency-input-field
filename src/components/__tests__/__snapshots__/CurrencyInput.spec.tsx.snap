// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CurrencyInput/> > Renders with default value 1`] = `
<input
  inputmode="decimal"
  type="text"
/>
`;

exports[`<CurrencyInput/> > Renders without error 1`] = `
<input
  class="form-control"
  id="validationCustom01"
  inputmode="decimal"
  name="inputName"
  placeholder="Â£1,000"
  type="text"
/>
`;
